odl/contrib/pytorch/layer.py

```diff
@@ -0,0 +1,232 @@
+        ----------
+        grad_output : `torch.tensor._TensorBase`
+            Tensor to which the Jacobian should be applied. See Notes
+            for details.
+
+        Returns
+        -------
+        result : `torch.autograd.variable.Variable`
+            Variable holding the result of applying the Jacobian to
+            ``grad_output``. See Notes for details.
+
+        Examples
+        --------
+        Compute the Jacobian adjoint of the matrix operator, which is the
+        operator of the transposed matrix. We compose with the ``sum``
+        functional to be able to evaluate ``grad``:
+
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> odl_op = odl.MatrixOperator(matrix)
+        >>> torch_op = TorchOperator(odl_op)
+        >>> x = torch.DoubleTensor([1, 2, 3])
+        >>> x_var = torch.autograd.Variable(x, requires_grad=True)
+        >>> op_x_var = torch_op(x_var)
+        >>> cost = op_x_var.sum()
+        >>> cost.backward()
+        >>> x_var.grad  # should be matrix.T.dot([1, 1])
+        Variable containing:
+         1
+         1
+         2
+        [torch.DoubleTensor of size 3]
+
+        Compute the gradient of a custom functional:
+
+        >>> odl_func = odl.solvers.L2NormSquared(odl.rn(3))
+        >>> torch_func = TorchOperator(odl_func)
+        >>> x = torch.DoubleTensor([1, 2, 3])
+        >>> x_var = torch.autograd.Variable(x, requires_grad=True)
+        >>> func_x_var = torch_func(x_var)
+        >>> func_x_var
+        Variable containing:
+         14
+        [torch.DoubleTensor of size 1]
+
+        >>> func_x_var.backward()
+        >>> x_var.grad  # Should be 2 * x
+        Variable containing:
+         2
+         4
+         6
+        [torch.DoubleTensor of size 3]
+
+        Notes
+        -----
+        This method applies the contribution of this node, i.e., the
+        transpose of the Jacobian of its outputs with respect to its inputs,
+        to the gradients of some cost function with respect to the outputs
+        of this node.
+
+        Example: Assume that this node computes :math:`x \mapsto C(f(x))`,
+        where :math:`x` is a tensor variable and :math:`C` is a scalar-valued
+        function. In ODL language, what ``backward`` should compute is
+
+        .. math::
+            \\nabla(C \circ f)(x) = f'(x)^*\\big(\\nabla C (f(x))\\big)
+
+        according to the chain rule. In ODL code, this corresponds to ::
+
+            f.derivative(x).adjoint(C.gradient(f(x))).
+
+        Hence, the parameter ``grad_output`` is a tensor variable containing
+        :math:`y = \\nabla C(f(x))`. Then, ``backward`` boils down to
+        computing ``[f'(x)^*(y)]`` using the input ``x`` stored during
+        the previous `forward` pass.
+        """
+        if not self.operator.is_linear:
+            input_arr = self.saved_variables[0].data.numpy()
+            if any(s == 0 for s in input_arr.strides):
```

Comment:
Does it not suffice that 2 of them are zero for us to get issues?

Reply:
Already 1 zero is too much, at least potentially. I'd rather stay away from those guys completely (note the `any`).

Reply:
Ah, you are correct.